[lldb] Clear ModuleList shared modules in SBDebugger::Clear

asavonic opened the pull request:
Shared modules are stored in a global `ModuleList` cache, and it is intentionally leaked to avoid doing cleanup when lldb exits.

However, when lldb is used as a library, we need a way to manage opened modules to avoid problems with file locks (on some systems) for modules that we no longer need.
It is possible to use `SBDebugger::MemoryPressureDetected` (which calls `ModuleList::RemoveOrphanSharedModules`) to clear all modules with `use_count == 1`, but this has to be done after `SBDebugger::Destroy`, because it may keep pointers to some modules.

The patch adds the following:

  - If enabled, `SBDebugger::Destroy` now calls `ModuleList::RemoveOrphanSharedModules` after all its pointers are released.
    Note that if the program keeps pointers to modules elsewhere, they will not be released, and they will not be removed from the cache.
    When a complete tear down is required, the program must ensure that all pointers are released before calling `SBDebugger::Destroy`.

  - `SBDebugger::SetClearSharedModules(bool)` API to enable this behavior.

  - `settings set clear-shared-modules true` command line option that is equivalent to `SetClearSharedModules(true)`.

This new behaviour is turned off by default: the debugger does not release these shared modules until the process exits, or until `SBDebugger::MemoryPressureDetected`.

The option provides a clear way for library users to communicate that complete cleanup is required, without relying on side-effects of `MemoryPressureDetected`.
It also allows to evaluate impact of clearing modules on performance of standalone debuggers.

asavonic committed:
[lldb] Clear ModuleList shared modules in SBDebugger::Clear

Shared modules are stored in a global ModuleList cache, and it is
intentionally leaked to avoid doing cleanup when lldb exits.

However, when lldb is used as a library, we need a way to manage
opened modules to avoid problems with file locks (on some systems) for
modules that we no longer need.

It should be possible to record all loaded modules and use
ModuleList::RemoveSharedModule and RemoveOrphanSharedModules functions
to clear the cache, but these functions are not available in the API.
This approach is also way too complicated when we just need to cleanup
the library.

The patch adds the following:

  - ModuleList::ClearSharedModules function to clear all shared
    modules.

  - SBDebugger::SetClearSharedModules(bool) API to enable clearing
    during SBDebugger::Clear().

  - `settings set clear-shared-modules true` command line option
    that is equivalent to SetClearSharedModules(true).

This new behaviour is turned off by default: the debugger does not
release these shared modules until the process exits.

asavonic commented:
> > However, when lldb is used as a library, we need a way to manage opened modules to avoid problems with file locks (on some systems) for modules that we no longer need.
> 
> Can you give a concrete example? This seems like its own issue that should be solved rather than worked around. I wouldn't expect us to need to keep files open for as long as a module exists.

Right, it is a separate issue. I'll follow up later.

> 
> > It should be possible to record all loaded modules and use `ModuleList::RemoveSharedModule` and `RemoveOrphanSharedModules` functions to clear the cache, but these functions are not available in the API. This approach is also way too complicated when we just need to cleanup the library.
> 
> This is exposed from the SB API through `SBDebugger::MemoryPressureDetected`. Not sure what's "complicated" about calling his function, but I do agree that this shouldn't necessary be an explicit operation and would make a good default. The natural place to do this is in `SBDebugger::Destroy`. You can have many debuggers and it seems like a bad idea to clear the modules when you clear one of them.

How should a proper debugger shutdown sequence look like?

1. Release all `SBModule` pointers.
2. Call `SBDebugger::MemoryPressureDetected`.
3. Call `SBDebugger::Destroy`.

`SBDebugger::MemoryPressureDetected` only removes shared pointers to modules with `use_count == 1`. Assuming that the calling program releases all its modules in (1), we need to make sure that the debugger does not keep any pointers elsewhere before (2). Can we rely on this?

If any module survives `MemoryPressureDetected`, they will not be released by `SBDebugger::Destroy`.

asavonic commented:
> The natural place to do this is in `SBDebugger::Destroy`. You can have many debuggers and it seems like a bad idea to clear the modules when you clear one of them.

Oh, right. If we call `MemoryPressureDetected` from `SBDebugger::Destroy`, we can make sure that all modules are released before `MemoryPressureDetected`.

JDevlieghere commented:
> How should a proper debugger shutdown sequence look like?
> 
> 1. Release all `SBModule` pointers.
> 2. Call `SBDebugger::MemoryPressureDetected`.
> 3. Call `SBDebugger::Destroy`.

Ideally, `SBDebugger::Destroy` would clean up all the modules in the module cache. You say we don't do that and that we leak them intentionally, which suggests that we don't do that on purpose. It'd be interesting to understand why, but I think that's fundamentally at odds with what you're trying to achieve. If there's a "good" reason and changing that is not an option, `SBDebugger::MemoryPressureDetected` could be an escape hatch. 

> `SBDebugger::MemoryPressureDetected` only removes shared pointers to modules with `use_count == 1`. Assuming that the calling program releases all its modules in (1), we need to make sure that the debugger does not keep any pointers elsewhere before (2). Can we rely on this?

Yes, if you destroyed all the debuggers and don't hold onto modules yourself, that should release all the modules. We actually test for this after every test in our test suite.

JDevlieghere commented:
> Oh, right. If we call `MemoryPressureDetected` from `SBDebugger::Destroy`, we can make sure that all modules are released before `MemoryPressureDetected`.

Aha, so it already kind-of works the way I was suggesting. Instead of forcefully deleting the modules, we ask nicely to release the ones that have no uses anymore. That seems very reasonable. I still wonder why we can't just boot the whole module cache but I'm sure there is a reason...

asavonic commented:
> You say we don't do that and that we leak them intentionally, which suggests that we don't do that on purpose. It'd be interesting to understand why, but I think that's fundamentally at odds with what you're trying to achieve.

The cache was added in 6f4d8af7 back in 2012. I guess It makes sense for a standalone debugger.

I'll update the patch to use `RemoveOrphanSharedModules` instead of `ClearSharedModules`. 
`SetClearSharedModules` option should probably stay to avoid performance regressions for standalone debuggers.

jimingham commented:
No one SBDebugger should force-clear the Shared Module Cache.  That's a resource shared among all the SBDebuggers, and no SBDebugger can know what the other debugger's intentions are w.r.t. it.  That seems wrong to me.  

The most an SBDebugger should do on Destroy or Clear is call MemoryPressureDetected or a similar lower-level API to clear all the modules that it is the only owner of.  It should not clear the cache out from under other users.

Even in its current reduced form I don't think this patch is a good idea.

asavonic committed:
Use RemoveOrphanSharedModules

asavonic committed:
Remove ClearSharedModules

asavonic commented:
Updated the patch to [use RemoveOrphanSharedModules](https://github.com/llvm/llvm-project/pull/147289/commits/7a27fe962726c3190ce6ad7d635e3ae30a0f52f7) instead of force-clearing the cache.
Changed the description as well.

> No one SBDebugger should force-clear the Shared Module Cache. That's a resource shared among all the SBDebuggers, and no SBDebugger can know what the other debugger's intentions are w.r.t. it. That seems wrong to me.

Agree.

> The most an SBDebugger should do on Destroy or Clear is call MemoryPressureDetected or a similar lower-level API to clear all the modules that it is the only owner of. It should not clear the cache out from under other users.

The last revision should do that. The downside is that the program must release all shared pointers to modules before calling `Destroy`, or they leak again.

JDevlieghere reviewed (changes requested):
This patch is still trying attempting to clear the module list when clearing the debugger. As Jim said, that doesn't make sense. The module list is global and spans multiple debuggers. 

Maybe I misunderstand the problem, but AFAIK all you need to achieve what you're trying to do is:

```
diff --git a/lldb/source/Core/Debugger.cpp b/lldb/source/Core/Debugger.cpp
index ed674ee1275c..c6390e90f726 100644
--- a/lldb/source/Core/Debugger.cpp
+++ b/lldb/source/Core/Debugger.cpp
@@ -742,6 +742,8 @@ void Debugger::Terminate() {
       g_debugger_list_ptr->clear();
     }
   }
+
+  ModuleList::RemoveOrphanSharedModules(/*mandatory=*/true);
 }
 
 void Debugger::SettingsInitialize() { Target::SettingsInitialize(); }
```

This is going to remove all the orphaned modules (= all modules if you're not holding on to any shared pointers, which is by design) when you terminate LLDB (i.e. the last thing you do with LLDB).

JDevlieghere commented:
> Do we need to tie this to any option, or we can clear these modules unconditionally? The current patch uses a property, but I don't see how to do this with `static Debugger::Terminate`.

I don't think this needs to be an option if we go with the approach of only removing orphaned modules.

jimingham commented:
Any Debugger that wants to can call MemoryPressureDetected or RemoveOrphanedSharedModules.  That's always been true.  But no Debugger should delete modules out from under any other.  And appropriately there are no API's that actually do that.

When Debugger::Terminate is called we should go through and close anything remaining on the module list.  Note, we should have closed down all the SBDebuggers by the time we get to clearing the global module list in Terminate, so there really shouldn't be anything with a non-zero reference count on it.  We test that in the test suite at the end of each test, and except for a few bugs, that's generally true.  So the normal RemoveOrphanedSharedModules in Terminate should do the job, but since we're tearing down the library at that point, force removing the rest should be fine, and doesn't need to be controlled by a setting.

Note however that the `mandatory` in RemoveOrphanedSharedModules is a bit of a misleading parameter name.  It isn't a "force these modules to be deleted" parameter, it only governs whether getting the module list lock is mandatory for doing the clearing, or if we should just do a try_lock.  When we get to deleting modules, we still obey the reference count.

So if you want to really force that on exit there are NO modules in the shared module cache, I'd suggest adding another API like `RemoveAllSharedModules` that ignores reference count - but that API should only ever be called from Debugger::Terminate.

asavonic commented:
@jimingham, thanks a lot for the explanation! I think `MemoryPressureDetected` is a better solution than what I was trying to do originally. 

> I'd suggest adding another API like `RemoveAllSharedModules` that ignores reference count - but that API should only ever be called from Debugger::Terminate.

I almost prepared a patch to add that, but tests highlighted another problem - there seems to be no way to recover from `Debugger::Terminate` and re-initialize. Initialization fails because Terminate leaves global pointers as non-zero and we get `Debugger::Initialize called more than once!`. Without a way to re-initialize, this dance around modules in Terminate makes even less sense. I think we can drop this patch, I don't see a practical reason to do it.

Thank you @jimingham and @JDevlieghere for your guidance here.
